### PR TITLE
feat: support autoscaling on broker and bookie

### PR DESCRIPTION
**Pull request by ericsyh**
*(If this PR fixes a github issue, please add `Fixes #<xyz>`.)*

Fixes #<xyz>

*(or if this PR is one task of a github issue, please add `Master Issue: #<xyz>` to link to the master issue.)*

Master Issue: #<xyz>

### Motivation

*Explain here the context, and why you're making that change. What is the problem you're trying to solve.*

### Modifications

*Describe the modifications you've done.*

### Verifying this change

- [x] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (10MB)*
  - *Extended integration test for recovery after broker failure*

### Documentation

Check the box below.

Need to update docs? 

- [ ] `doc-required` 
  
  (If you need help on updating docs, create a doc issue)
  
- [ ] `no-need-doc` 
  
  (Please explain why)
  
- [x] `doc` 
  
  (If this PR contains doc changes)

### CODE REVIEW

**Comment by ericsyh** [comment]
### Example

```
bookkeeper:
  autoScaling:
    enabled: true
    minReplicas: 1
    maxReplicas: 4
    metrics:
     - pods:
        metric:
          name: cpu_usage
        target:
          averageValue: "75"
          type: AverageValue
      type: Pods

broker:
  autoScaling:
    enabled: true
    minReplicas: 1
    maxReplicas: 4
    metrics:
     - pods:
        metric:
          name: cpu_usage
        target:
          averageValue: "75"
          type: AverageValue
      type: Pods
```

File: charts/sn-platform/values.yaml [hunk 1]
```diff
@@ -1099,6 +1107,14 @@ broker:
     # annotations that will be added on the PulsarBroker CR only.
     annotations: {}
 
+  # The field autoScaling will track the average metrics usage of the Pulsar nodes and adjusts the nodes to keep at the target metrics usage level.
+  # If the average metrics usage for nodes is over the target, scaling controller will scale out to bring more nodes and if the average metrics for the nodes is less than the target, scaling controller will downscale nodes to save resources.
+  autoScaling:
+    enabled: false
+    minReplicas: 1
+    maxReplicas: 4
+    metrics: {}
```
Comment (ignacioli) [inline charts/sn-platform/values.yaml:1116]: what are the available metrics here?

Comment (ignacioli) [reply charts/sn-platform/values.yaml:1116]: It would be great if there is a list of the metric names that can be used for pulsar auto-scaling.

Comment (ericsyh) [reply charts/sn-platform/values.yaml:1116]: > what are the available metrics here?
It depends on the metrics backend service used. If users deploy the Metrics Server then only resource metrics could be used. But with prometheus and prometheus adapter, all metrics from the prometheus could be used by the auto-scaling.

Comment (ignacioli) [reply charts/sn-platform/values.yaml:1116]: The `cpu_usage` here is a  custom metric implemented by [prometheus-adapter](https://github.com/kubernetes-sigs/prometheus-adapter) other than the general prometheus metric,  the user should know the catalog of the supported metrics of broker autoScaling.

**Review by ignacioli — APPROVED** [review]

**Pull request merged**